File: src/main/java/villagecompute/calendar/services/TemplateService.java
package villagecompute.calendar.services;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Inject;
import jakarta.transaction.Transactional;
import org.jboss.logging.Logger;
import villagecompute.calendar.api.graphql.inputs.TemplateInput;
import villagecompute.calendar.data.models.CalendarTemplate;
import villagecompute.calendar.data.repositories.CalendarTemplateRepository;

import java.util.Optional;
import java.util.UUID;

/**
 * Service layer for calendar template management operations.
 * Handles template CRUD operations, validation, and preview image uploads.
 * All operations are admin-only.
 */
@ApplicationScoped
public class TemplateService {

    private static final Logger LOG = Logger.getLogger(TemplateService.class);
    private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();

    @Inject
    CalendarTemplateRepository templateRepository;

    @Inject
    StorageService storageService;

    /**
     * Create a new calendar template.
     * Validates the template configuration JSONB before creation.
     *
     * @param input Template creation data
     * @return Created template
     * @throws IllegalArgumentException if input validation fails
     */
    @Transactional
    public CalendarTemplate createTemplate(TemplateInput input) {
        LOG.infof("Creating template: name=%s", input.name);

        // Parse configuration JSON string
        JsonNode configNode = parseConfiguration(input.configuration);

        // Validate configuration structure
        validateTemplateConfiguration(configNode);

        // Check for duplicate name
        Optional<CalendarTemplate> existing = templateRepository.findByName(input.name);
        if (existing.isPresent()) {
            LOG.errorf("Template with name '%s' already exists", input.name);
            throw new IllegalArgumentException("Template with this name already exists");
        }

        // Create new template entity
        CalendarTemplate template = new CalendarTemplate();
        template.name = input.name;
        template.description = input.description;
        template.configuration = configNode;
        template.thumbnailUrl = input.thumbnailUrl;
        template.isActive = input.isActive != null ? input.isActive : true;
        template.isFeatured = input.isFeatured != null ? input.isFeatured : false;
        template.displayOrder = input.displayOrder != null ? input.displayOrder : 0;
        template.previewSvg = input.previewSvg;

        // Persist
        template.persist();

        LOG.infof("Created template: ID=%s, name=%s", template.id, template.name);

        return template;
    }

    /**
     * Update an existing calendar template.
     * Validates configuration if provided.
     *
     * @param id Template ID
     * @param input Template update data
     * @return Updated template
     * @throws IllegalArgumentException if template not found or validation fails
     */
    @Transactional
    public CalendarTemplate updateTemplate(UUID id, TemplateInput input) {
        LOG.infof("Updating template: ID=%s", id);

        // Find existing template
        CalendarTemplate template = CalendarTemplate.<CalendarTemplate>findByIdOptional(id).orElse(null);
        if (template == null) {
            LOG.errorf("Template not found: %s", id);
            throw new IllegalArgumentException("Template not found");
        }

        // Parse and validate configuration if provided
        JsonNode configNode = null;
        if (input.configuration != null && !input.configuration.isBlank()) {
            configNode = parseConfiguration(input.configuration);
            validateTemplateConfiguration(configNode);
        }

        // Check for name conflicts (if name is being changed)
        if (input.name != null && !input.name.equals(template.name)) {
            Optional<CalendarTemplate> existing = templateRepository.findByName(input.name);
            if (existing.isPresent() && !existing.get().id.equals(id)) {
                LOG.errorf("Template with name '%s' already exists", input.name);
                throw new IllegalArgumentException("Template with this name already exists");
            }
        }

        // Apply updates
        if (input.name != null) {
            template.name = input.name;
        }
        if (input.description != null) {
            template.description = input.description;
        }
        if (configNode != null) {
            template.configuration = configNode;
        }
        if (input.thumbnailUrl != null) {
            template.thumbnailUrl = input.thumbnailUrl;
        }
        if (input.isActive != null) {
            template.isActive = input.isActive;
        }
        if (input.isFeatured != null) {
            template.isFeatured = input.isFeatured;
        }
        if (input.displayOrder != null) {
            template.displayOrder = input.displayOrder;
        }
        if (input.previewSvg != null) {
            template.previewSvg = input.previewSvg;
        }

        // Persist changes
        template.persist();

        LOG.infof("Updated template: ID=%s, name=%s", template.id, template.name);

        return template;
    }

    /**
     * Delete a calendar template using soft delete.
     * Sets isActive=false instead of permanently removing the record.
     * Templates with existing calendars can be soft-deleted to preserve data integrity.
     *
     * @param id Template ID
     * @throws IllegalArgumentException if template not found
     */
    @Transactional
    public void deleteTemplate(UUID id) {
        LOG.infof("Soft-deleting template: ID=%s", id);

        // Find existing template
        CalendarTemplate template = CalendarTemplate.<CalendarTemplate>findByIdOptional(id).orElse(null);
        if (template == null) {
            LOG.errorf("Template not found: %s", id);
            throw new IllegalArgumentException("Template not found");
        }

        // Soft delete by setting isActive=false
        template.isActive = false;
        template.persist();

        LOG.infof("Soft-deleted template: ID=%s, name=%s, isActive=false", id, template.name);
    }

    /**
     * Upload a preview image for a template to R2 storage.
     * Updates the template's thumbnailUrl field with the public URL.
     *
     * @param templateId Template ID
     * @param imageBytes Image file bytes
     * @param contentType MIME type (e.g., "image/png", "image/jpeg")
     * @return Public URL of the uploaded image
     * @throws IllegalArgumentException if template not found
     */
    @Transactional
    public String uploadPreviewImage(UUID templateId, byte[] imageBytes, String contentType) {
        LOG.infof("Uploading preview image for template: ID=%s", templateId);

        // Find existing template
        CalendarTemplate template = CalendarTemplate.<CalendarTemplate>findByIdOptional(templateId).orElse(null);
        if (template == null) {
            LOG.errorf("Template not found: %s", templateId);
            throw new IllegalArgumentException("Template not found");
        }

        // Generate unique filename
        String filename = "template-" + templateId + "-" + System.currentTimeMillis() + getFileExtension(contentType);

        // Upload to R2
        String publicUrl = storageService.uploadFile(filename, imageBytes, contentType);

        // Update template with new thumbnail URL
        template.thumbnailUrl = publicUrl;
        template.persist();

        LOG.infof("Uploaded preview image for template %s: %s", templateId, publicUrl);

        return publicUrl;
    }

    /**
     * Parse a JSON string into a JsonNode.
     *
     * @param configurationJson JSON string to parse
     * @return Parsed JsonNode
     * @throws IllegalArgumentException if JSON is invalid
     */
    private JsonNode parseConfiguration(String configurationJson) {
        if (configurationJson == null || configurationJson.isBlank()) {
            throw new IllegalArgumentException("Configuration cannot be null or empty");
        }

        try {
            return OBJECT_MAPPER.readTree(configurationJson);
        } catch (JsonProcessingException e) {
            LOG.errorf("Invalid JSON in configuration: %s", e.getMessage());
            throw new IllegalArgumentException("Invalid JSON in configuration: " + e.getMessage(), e);
        }
    }

    /**
     * Validate the structure of a template configuration JSONB.
     * Ensures the configuration is a valid JSON object.
     *
     * @param configuration JsonNode representing the template configuration
     * @throws IllegalArgumentException if configuration is invalid
     */
    private void validateTemplateConfiguration(JsonNode configuration) {
        if (configuration == null || configuration.isNull()) {
            throw new IllegalArgumentException("Template configuration cannot be null");
        }

        if (!configuration.isObject()) {
            throw new IllegalArgumentException("Template configuration must be a JSON object");
        }

        LOG.debug("Template configuration validation passed");
    }

    /**
     * Get file extension from content type.
     *
     * @param contentType MIME type
     * @return File extension with dot (e.g., ".png")
     */
    private String getFileExtension(String contentType) {
        if (contentType == null) {
            return "";
        }

        return switch (contentType.toLowerCase()) {
            case "image/png" -> ".png";
            case "image/jpeg", "image/jpg" -> ".jpg";
            case "image/gif" -> ".gif";
            case "image/webp" -> ".webp";
            case "image/svg+xml" -> ".svg";
            default -> "";
        };
    }
}

File: src/main/webui/src/services/templateService.ts
/**
 * Template Service
 * Provides GraphQL API calls for template CRUD operations
 */

import type { CalendarTemplate, TemplateInput } from '../types/template'

const GRAPHQL_ENDPOINT = '/graphql'

interface GraphQLResponse<T> {
  data?: T
  errors?: Array<{ message: string }>
}

interface ServiceResponse<T> {
  data: T | null
  error: string | null
}

/**
 * Get JWT token from localStorage
 */
function getAuthToken(): string | null {
  return localStorage.getItem('authToken')
}

/**
 * Make GraphQL request with authorization
 */
async function graphqlFetch<T>(
  query: string,
  variables?: Record<string, any>
): Promise<ServiceResponse<T>> {
  try {
    const token = getAuthToken()
    const headers: Record<string, string> = {
      'Content-Type': 'application/json',
    }

    if (token) {
      headers['Authorization'] = `Bearer ${token}`
    }

    const response = await fetch(GRAPHQL_ENDPOINT, {
      method: 'POST',
      headers,
      body: JSON.stringify({ query, variables }),
    })

    if (!response.ok) {
      if (response.status === 401) {
        return { data: null, error: 'Unauthorized. Please log in.' }
      }
      if (response.status === 403) {
        return { data: null, error: 'Forbidden. Admin access required.' }
      }
      return { data: null, error: `HTTP error: ${response.status}` }
    }

    const result: GraphQLResponse<T> = await response.json()

    if (result.errors && result.errors.length > 0) {
      return { data: null, error: result.errors[0].message }
    }

    if (!result.data) {
      return { data: null, error: 'No data returned from server' }
    }

    return { data: result.data, error: null }
  } catch (err) {
    return { data: null, error: err instanceof Error ? err.message : 'Unknown error' }
  }
}

/**
 * Fetch all templates (includes inactive for admins)
 */
export async function fetchTemplates(
  isActive?: boolean
): Promise<ServiceResponse<CalendarTemplate[]>> {
  const query = `
    query Templates($isActive: Boolean) {
      templates(isActive: $isActive) {
        id
        name
        description
        configuration
        thumbnailUrl
        previewSvg
        isActive
        isFeatured
        displayOrder
        created
        updated
      }
    }
  `

  const result = await graphqlFetch<{ templates: CalendarTemplate[] }>(query, { isActive })

  if (result.error) {
    return { data: null, error: result.error }
  }

  return { data: result.data?.templates || [], error: null }
}

/**
 * Fetch single template by ID
 */
export async function fetchTemplate(id: string): Promise<ServiceResponse<CalendarTemplate>> {
  const query = `
    query Template($id: ID!) {
      template(id: $id) {
        id
        name
        description
        configuration
        thumbnailUrl
        previewSvg
        isActive
        isFeatured
        displayOrder
        created
        updated
      }
    }
  `

  const result = await graphqlFetch<{ template: CalendarTemplate }>(query, { id })

  if (result.error) {
    return { data: null, error: result.error }
  }

  if (!result.data?.template) {
    return { data: null, error: 'Template not found' }
  }

  return { data: result.data.template, error: null }
}

/**
 * Create new template (admin only)
 */
export async function createTemplate(
  input: TemplateInput
): Promise<ServiceResponse<CalendarTemplate>> {
  const query = `
    mutation CreateTemplate($input: TemplateInput!) {
      createTemplate(input: $input) {
        id
        name
        description
        configuration
        thumbnailUrl
        previewSvg
        isActive
        isFeatured
        displayOrder
        created
        updated
      }
    }
  `

  const result = await graphqlFetch<{ createTemplate: CalendarTemplate }>(query, { input })

  if (result.error) {
    return { data: null, error: result.error }
  }

  if (!result.data?.createTemplate) {
    return { data: null, error: 'Failed to create template' }
  }

  return { data: result.data.createTemplate, error: null }
}

/**
 * Update existing template (admin only)
 */
export async function updateTemplate(
  id: string,
  input: TemplateInput
): Promise<ServiceResponse<CalendarTemplate>> {
  const query = `
    mutation UpdateTemplate($id: ID!, $input: TemplateInput!) {
      updateTemplate(id: $id, input: $input) {
        id
        name
        description
        configuration
        thumbnailUrl
        previewSvg
        isActive
        isFeatured
        displayOrder
        created
        updated
      }
    }
  `

  const result = await graphqlFetch<{ updateTemplate: CalendarTemplate }>(query, { id, input })

  if (result.error) {
    return { data: null, error: result.error }
  }

  if (!result.data?.updateTemplate) {
    return { data: null, error: 'Failed to update template' }
  }

  return { data: result.data.updateTemplate, error: null }
}

/**
 * Delete template (admin only)
 */
export async function deleteTemplate(id: string): Promise<ServiceResponse<boolean>> {
  const query = `
    mutation DeleteTemplate($id: ID!) {
      deleteTemplate(id: $id)
    }
  `

  const result = await graphqlFetch<{ deleteTemplate: boolean }>(query, { id })

  if (result.error) {
    return { data: null, error: result.error }
  }

  return { data: result.data?.deleteTemplate || false, error: null }
}
